
<!doctype html>
<html lang="zh" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
        <meta name="author" content="biolxy">
      
      
        <link rel="canonical" href="https://biolxy.github.io/blog/category/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6.html">
      
      
        <link rel="prev" href="%E7%94%9F%E6%B4%BB%E6%9C%AD%E8%AE%B0.html">
      
      
        <link rel="next" href="%E7%BB%9F%E8%AE%A1%E5%AD%A6.html">
      
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.14">
    
    
      
        <title>生物信息学 - biolxy的知识库</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.342714a4.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
        <script src="https://unpkg.com/iframe-worker/shim"></script>
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="orange">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#_1" class="md-skip">
          跳转至
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="页眉">
    <a href="../../index.html" title="biolxy的知识库" class="md-header__button md-logo" aria-label="biolxy的知识库" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            biolxy的知识库
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              生物信息学
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="orange"  aria-label="切换至夜间模式"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="切换至夜间模式" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m17.75 4.09-2.53 1.94.91 3.06-2.63-1.81-2.63 1.81.91-3.06-2.53-1.94L12.44 4l1.06-3 1.06 3zm3.5 6.91-1.64 1.25.59 1.98-1.7-1.17-1.7 1.17.59-1.98L15.75 11l2.06-.05L18.5 9l.69 1.95zm-2.28 4.95c.83-.08 1.72 1.1 1.19 1.85-.32.45-.66.87-1.08 1.27C15.17 23 8.84 23 4.94 19.07c-3.91-3.9-3.91-10.24 0-14.14.4-.4.82-.76 1.27-1.08.75-.53 1.93.36 1.85 1.19-.27 2.86.69 5.83 2.89 8.02a9.96 9.96 0 0 0 8.02 2.89m-1.64 2.02a12.08 12.08 0 0 1-7.8-3.47c-2.17-2.19-3.33-5-3.49-7.82-2.81 3.14-2.7 7.96.31 10.98 3.02 3.01 7.84 3.12 10.98.31"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="indigo" data-md-color-accent="orange"  aria-label="切换至日间模式"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="切换至日间模式" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 7a5 5 0 0 1 5 5 5 5 0 0 1-5 5 5 5 0 0 1-5-5 5 5 0 0 1 5-5m0 2a3 3 0 0 0-3 3 3 3 0 0 0 3 3 3 3 0 0 0 3-3 3 3 0 0 0-3-3m0-7 2.39 3.42C13.65 5.15 12.84 5 12 5s-1.65.15-2.39.42zM3.34 7l4.16-.35A7.2 7.2 0 0 0 5.94 8.5c-.44.74-.69 1.5-.83 2.29zm.02 10 1.76-3.77a7.131 7.131 0 0 0 2.38 4.14zM20.65 7l-1.77 3.79a7.02 7.02 0 0 0-2.38-4.15zm-.01 10-4.14.36c.59-.51 1.12-1.14 1.54-1.86.42-.73.69-1.5.83-2.29zM12 22l-2.41-3.44c.74.27 1.55.44 2.41.44.82 0 1.63-.17 2.37-.44z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="搜索" placeholder="搜索" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="查找">
        
        <button type="reset" class="md-search__icon md-icon" title="清空当前内容" aria-label="清空当前内容" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            正在初始化搜索引擎
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="标签" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../../index.html" class="md-tabs__link">
        
  
  
    
  
  Home

      </a>
    </li>
  

      
        
  
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../index.html" class="md-tabs__link">
          
  
  
    
  
  Blog

        </a>
      </li>
    
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../NGS/0.Background-knowledge.html" class="md-tabs__link">
          
  
  
    
  
  肿瘤基因检测

        </a>
      </li>
    
  

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="导航栏" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../index.html" title="biolxy的知识库" class="md-nav__button md-logo" aria-label="biolxy的知识库" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    biolxy的知识库
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../index.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
      
        
          
        
      
        
      
        
      
    
    
    
      
        
        
      
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" checked>
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../index.html" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    Blog
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_2" id="__nav_2_label" tabindex="">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            Blog
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_2_2" >
        
          
          <label class="md-nav__link" for="__nav_2_2" id="__nav_2_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    归档
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_2_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2_2">
            <span class="md-nav__icon md-icon"></span>
            归档
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    
  
  
    <li class="md-nav__item">
      <a href="../archive/2025.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2025
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    
  
  
    <li class="md-nav__item">
      <a href="../archive/2024.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2024
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    
  
  
    <li class="md-nav__item">
      <a href="../archive/2022.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2022
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    
  
  
    <li class="md-nav__item">
      <a href="../archive/2021.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2021
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    
  
  
    <li class="md-nav__item">
      <a href="../archive/2020.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2020
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    
  
  
    <li class="md-nav__item">
      <a href="../archive/2019.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2019
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    
  
  
    <li class="md-nav__item">
      <a href="../archive/2018.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2018
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
    
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2_3" checked>
        
          
          <label class="md-nav__link" for="__nav_2_3" id="__nav_2_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    分类
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_2_3_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_2_3">
            <span class="md-nav__icon md-icon"></span>
            分类
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    
  
  
    <li class="md-nav__item">
      <a href="cc.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    C/C++
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    
  
  
    <li class="md-nav__item">
      <a href="conda.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Conda
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    
  
  
    <li class="md-nav__item">
      <a href="docker.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Docker
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    
  
  
    <li class="md-nav__item">
      <a href="linux.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Linux
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    
  
  
    <li class="md-nav__item">
      <a href="python.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Python
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    
  
  
    <li class="md-nav__item">
      <a href="r.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    R
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    
  
  
    <li class="md-nav__item">
      <a href="rusedest.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    RuseDest
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    
  
  
    <li class="md-nav__item">
      <a href="windows.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Windows
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    
  
  
    <li class="md-nav__item">
      <a href="develop.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    develop
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    
  
  
    <li class="md-nav__item">
      <a href="%E5%8D%9A%E5%AE%A2.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    博客
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    
  
  
    <li class="md-nav__item">
      <a href="%E7%94%9F%E6%B4%BB%E6%9C%AD%E8%AE%B0.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    生活札记
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
    
      
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    生物信息学
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6.html" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    生物信息学
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="目录">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      目录
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#panel-bed" class="md-nav__link">
    <span class="md-ellipsis">
      计算panel bed编码区长度
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#_1" class="md-nav__link">
    <span class="md-ellipsis">
      准确度和灵敏度和特异性
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#_1" class="md-nav__link">
    <span class="md-ellipsis">
      生物信息学习技术路线(生信指月录)
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#_1" class="md-nav__link">
    <span class="md-ellipsis">
      获取启动子序列
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#reads-count-tpm-fpkm" class="md-nav__link">
    <span class="md-ellipsis">
      由 reads count 数计算 TPM 和FPKM
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#linux" class="md-nav__link">
    <span class="md-ellipsis">
      生物信息学常用的Linux命令
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
    
  
  
    <li class="md-nav__item">
      <a href="%E7%BB%9F%E8%AE%A1%E5%AD%A6.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    统计学
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    
  
  
    <li class="md-nav__item">
      <a href="%E8%82%BF%E7%98%A4%E5%9F%BA%E5%9B%A0%E7%BB%84.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    肿瘤基因组
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_3" >
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    肿瘤基因检测
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            肿瘤基因检测
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../NGS/0.Background-knowledge.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    0.背景知识 Background knowledge
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../NGS/1.reads-align.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    1.序列比对 reads align
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../NGS/2.remove-mark-dup-reads.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2.序列去重 mark/remove dup reads
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../NGS/3.fastp-bam-qc.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    3.质控 QC
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../NGS/4.somatic-mutation-calling.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    4.变异检测 somatic mutation calling
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../NGS/5.annotate.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    5.注释 annotate
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../NGS/6.signature.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    6.signature
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../NGS/7.Fusion.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    7.Fusion 基因融合
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../NGS/8.MSI.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    8.MSI 微卫星不稳定
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../NGS/9.TMB.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    9.TMB
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../NGS/10.HLAtype.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    10.HLAtype
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../NGS/11.Neoantigen.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    11.Neoantigen
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../NGS/12.HLALOH.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    12.HLALOH
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../NGS/13.tumor-abundance-clone.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    13.tumor abundance clone
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../NGS/14.HRD.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    14.HRD
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../NGS/5.MRD.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    15.ctDNA MRD
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../NGS/5.RNA.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    16.RNA
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../NGS/17.CNV.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    17.CNV
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../NGS/18.conta.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    18.污染判定
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../NGS/19.Medical-knowledge.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    19.医学知识库
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../NGS/20.report.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    20.自动化报告
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../NGS/21.workflow.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    21.workflow
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../NGS/22.data.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    22.数据传输存储
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../NGS/23.gender.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    23.性别判定
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../NGS/24.pv.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    24.性能验证
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../NGS/25.industry.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    25.行业相关
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="目录">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      目录
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#panel-bed" class="md-nav__link">
    <span class="md-ellipsis">
      计算panel bed编码区长度
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#_1" class="md-nav__link">
    <span class="md-ellipsis">
      准确度和灵敏度和特异性
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#_1" class="md-nav__link">
    <span class="md-ellipsis">
      生物信息学习技术路线(生信指月录)
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#_1" class="md-nav__link">
    <span class="md-ellipsis">
      获取启动子序列
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#reads-count-tpm-fpkm" class="md-nav__link">
    <span class="md-ellipsis">
      由 reads count 数计算 TPM 和FPKM
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#linux" class="md-nav__link">
    <span class="md-ellipsis">
      生物信息学常用的Linux命令
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
  <div class="md-content" data-md-component="content">
    <div class="md-content__inner">
      <header class="md-typeset">
        <h1 id="_1">生物信息学<a class="headerlink" href="#_1" title="Permanent link">&para;</a></h1>
      </header>
      
        <article class="md-post md-post--excerpt">
  <header class="md-post__header">
    
      <nav class="md-post__authors md-typeset">
        
          <span class="md-author">
            <img src="https://gravatar.com/avatar/d63d3425433de59d2c3e901977a581c7?size=256&cache=1715411947733" alt="biolxy">
          </span>
        
      </nav>
    
    <div class="md-post__meta md-meta">
      <ul class="md-meta__list">
        <li class="md-meta__item">
          <time datetime="2024-07-22 00:00:00+00:00">2024年7月22日星期一</time></li>
        
          <li class="md-meta__item">
            分类于
            
              <a href="%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6.html" class="md-meta__link">生物信息学</a></li>
        
        
          
          <li class="md-meta__item">
            
              需要 1 分钟阅读时间
            
          </li>
        
      </ul>
      
    </div>
  </header>
  <div class="md-post__content md-typeset">
    <h2 id="panel-bed"><a class="toclink" href="../2024/07/22/%E8%AE%A1%E7%AE%97panel-bed%E7%BC%96%E7%A0%81%E5%8C%BA%E9%95%BF%E5%BA%A6.html"><strong>计算panel bed编码区长度</strong></a></h2>
<div class="language-bash highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">Bash</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span></pre></div></td><td class="code"><div><pre><span></span><code><span id="__span-2-1"><span class="ch">#!/bin/bash</span>
</span><span id="__span-2-2"><span class="c1">#@File    :   run.sh</span>
</span><span id="__span-2-3"><span class="c1">#@Time    :   2023/11/21 09:32:17</span>
</span><span id="__span-2-4"><span class="c1">#@Author  :   biolxy</span>
</span><span id="__span-2-5"><span class="c1">#@Version :   1.0</span>
</span><span id="__span-2-6"><span class="c1">#@Contact :   biolxy@aliyun.com</span>
</span><span id="__span-2-7"><span class="c1">#@Desc    :   None</span>
</span><span id="__span-2-8">
</span><span id="__span-2-9"><span class="nv">inputbed</span><span class="o">=</span><span class="nv">$1</span>
</span><span id="__span-2-10">
</span><span id="__span-2-11"><span class="c1"># export PATH=/data/biogonco/lixy/bedtools/bedtools2/bin:$PATH  # v2.30.1 不行</span>
</span><span id="__span-2-12"><span class="nb">export</span><span class="w"> </span><span class="nv">PATH</span><span class="o">=</span>/data/bioinfo_project/bioinfo_miniconda3/bin:<span class="nv">$PATH</span><span class="w">  </span><span class="c1"># v2.30.0 不行</span>
</span><span id="__span-2-13"><span class="c1"># /data/biogonco/lixy/bedtools/bedtools2/bin/bedtools</span>
</span><span id="__span-2-14">
</span><span id="__span-2-15">bedtools<span class="w"> </span>merge<span class="w"> </span>-i<span class="w"> </span><span class="si">${</span><span class="nv">inputbed</span><span class="si">}</span><span class="w"> </span>&gt;<span class="w"> </span>merge.bed
</span><span id="__span-2-16">
</span><span id="__span-2-17">bedtools<span class="w"> </span>summary<span class="w"> </span>-i<span class="w"> </span>merge.bed<span class="w"> </span>-g<span class="w"> </span>/mnt/nas_001/project/oncodemo/genome_and_annotation/hg19/gatk/b37/chrom.sizes<span class="w"> </span>&gt;<span class="w"> </span>merge.bed.summary<span class="w">  </span><span class="c1"># 统计的是 bed 区间的长度</span>
</span><span id="__span-2-18">
</span><span id="__span-2-19">
</span><span id="__span-2-20"><span class="c1"># 去 gencode 下载 gencode.v41.annotation.gff3</span>
</span><span id="__span-2-21"><span class="c1"># wget https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_44/GRCh37_mapping/gencode.v44lift37.annotation.gff3.gz</span>
</span><span id="__span-2-22"><span class="c1"># /mnt/nas_001/project/oncodemo/genome_and_annotation/genome/gffread-0.12.7.Linux_x86_64/gffread gencode.v44lift37.annotation.gff3 -T -o gencode.v44lift37.annotation.bed</span>
</span><span id="__span-2-23"><span class="c1"># awk &#39;$3 == &quot;CDS&quot;&#39; gencode.v44lift37.annotation.bed | sed &#39;s/^chr//g&#39; &gt; coding_CDS.bed</span>
</span><span id="__span-2-24">
</span><span id="__span-2-25"><span class="c1"># coding_CDS.bed 为编码区bed</span>
</span><span id="__span-2-26">
</span><span id="__span-2-27">bedtools<span class="w"> </span>intersect<span class="w"> </span>-a<span class="w"> </span>merge.bed<span class="w"> </span>-b<span class="w"> </span>/mnt/nas_001/project/oncodemo/genome_and_annotation/genome/coding_CDS.bed<span class="w"> </span><span class="p">|</span><span class="w"> </span>sort<span class="w"> </span>-k1,1<span class="w"> </span>-k2,2n<span class="w"> </span><span class="p">|</span><span class="w">  </span>bedtools<span class="w"> </span>merge<span class="w"> </span>-i<span class="w"> </span>-<span class="w"> </span>&gt;<span class="w"> </span>panel.cds.bed
</span><span id="__span-2-28">
</span><span id="__span-2-29"><span class="c1"># panel.cds.bed 为 pane 的 编码区 bed</span>
</span><span id="__span-2-30">bedtools<span class="w"> </span>summary<span class="w"> </span>-i<span class="w"> </span>panel.cds.bed<span class="w"> </span>-g<span class="w"> </span>/mnt/nas_001/project/oncodemo/genome_and_annotation/hg19/gatk/b37/chrom.sizes<span class="w"> </span>&gt;<span class="w"> </span>panel.cds.summary<span class="w"> </span>
</span></code></pre></div></td></tr></table></div>
<h3 id="_1"><a class="toclink" href="../2024/07/22/%E8%AE%A1%E7%AE%97panel-bed%E7%BC%96%E7%A0%81%E5%8C%BA%E9%95%BF%E5%BA%A6.html#_1">注意</a></h3>
<p>要注意bedtools的版本问题，我测试的 v2.30.0可以，v2.30.1 不行</p>
    
  </div>
</article>
      
        <article class="md-post md-post--excerpt">
  <header class="md-post__header">
    
      <nav class="md-post__authors md-typeset">
        
          <span class="md-author">
            <img src="https://gravatar.com/avatar/d63d3425433de59d2c3e901977a581c7?size=256&cache=1715411947733" alt="biolxy">
          </span>
        
      </nav>
    
    <div class="md-post__meta md-meta">
      <ul class="md-meta__list">
        <li class="md-meta__item">
          <time datetime="2024-05-27 00:00:00+00:00">2024年5月27日星期一</time></li>
        
          <li class="md-meta__item">
            分类于
            
              <a href="%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6.html" class="md-meta__link">生物信息学</a></li>
        
        
          
          <li class="md-meta__item">
            
              需要 1 分钟阅读时间
            
          </li>
        
      </ul>
      
    </div>
  </header>
  <div class="md-post__content md-typeset">
    <h2 id="_1"><a class="toclink" href="../2024/05/27/%E5%87%86%E7%A1%AE%E5%BA%A6%E5%92%8C%E7%81%B5%E6%95%8F%E5%BA%A6%E5%92%8C%E7%89%B9%E5%BC%82%E6%80%A7.html">准确度和灵敏度和特异性</a></h2>
<h3 id="_2"><a class="toclink" href="../2024/05/27/%E5%87%86%E7%A1%AE%E5%BA%A6%E5%92%8C%E7%81%B5%E6%95%8F%E5%BA%A6%E5%92%8C%E7%89%B9%E5%BC%82%E6%80%A7.html#_2">定义</a></h3>
<p>Precision 准确度：指找出来的突变中的真阳性有多少 </p>
<p>Recall    召回率：指总突变集合有多少可以被找出来</p>
<h3 id="_3"><a class="toclink" href="../2024/05/27/%E5%87%86%E7%A1%AE%E5%BA%A6%E5%92%8C%E7%81%B5%E6%95%8F%E5%BA%A6%E5%92%8C%E7%89%B9%E5%BC%82%E6%80%A7.html#_3">公式</a></h3>
<p><strong>准确度（Precision）</strong>：</p>
<p>准确度是指在所有被预测为阳性（即突变）的样本中，实际为阳性（真阳性）的比例。换句话说，它衡量的是预测结果中真阳性的比例。其计算公式为：</p>
<p><span class="arithmatex">\(\text{Precision} = \frac{TP}{TP + FP}\)</span></p>
<p>其中，TP是真阳性的数量，FP是假阳性的数量。</p>
<p><strong>召回率（Recall）或 灵敏度/敏感度（Sensitivity）</strong>：</p>
<p>召回率是指在所有实际为阳性的样本中，被正确预测为阳性（即真阳性）的比例。它衡量的是测试方法捕捉到所有实际阳性样本的能力。其计算公式为：</p>
<p><span class="arithmatex">\(\text{Recall} = \frac{TP}{TP + FN}\)</span></p>
<p>其中，TP是真阳性的数量，FN是假阴性的数量。</p>
<p>这两个指标通常用于评估分类模型的性能，特别是在医学和生物学研究中，它们帮助研究者理解一个测试或分析方法在识别阳性样本方面的准确性和完整性。在体细胞变异检测的背景下，这两个指标对于评估变异检测算法的性能至关重要。</p>
<p><strong>特异性（Specificity）</strong>：</p>
<p>特异性是指在实际没有某种疾病的人群中，诊断测试能够正确排除非患者的能力。它衡量了测试对非患者的"特异性"，即测试能够准确地排除非患者的能力。
特异性的计算公式为：</p>
<p>特异性 = 真阴性（True Negative）/（真阴性 + 假阳性（False Positive））</p>
<p><span class="arithmatex">\(\text{Specificity}= \frac{TN}{FP + TN}\)</span></p>
<p><strong>注意</strong>： 这里不使用 特异性 这一概念，原因是在NGS数据calling中，难以确定真阴性位点的数量，特异性无法计算。</p>
<p><strong>F1_score</strong> </p>
<p>F1_score，是评估分类模型性能的一种指标，特别是在二分类问题中。它结合了精确度（Precision）和召回率（Recall）两个指标来提供一个单一的评分，以衡量模型的整体性能。
F1_score 是精确度和召回率的调和平均数，公式为：</p>
<p><span class="arithmatex">\(F1 = 2 \times \frac{\text{Precision} \times \text{Recall}}{\text{Precision} + \text{Recall}}\)</span></p>
<p>F1_score 的值范围是0到1，1表示完美的精确度和召回率，0表示模型性能很差。</p>
<p>在处理不平衡的数据集时，F1_score 特别有用，因为它同时考虑了精确度和召回率，而不是仅仅关注其中一个。这使得 F1_score 成为一个在多种情况下都相对平衡的性能度量标准。</p>
    
  </div>
</article>
      
        <article class="md-post md-post--excerpt">
  <header class="md-post__header">
    
      <nav class="md-post__authors md-typeset">
        
          <span class="md-author">
            <img src="https://gravatar.com/avatar/d63d3425433de59d2c3e901977a581c7?size=256&cache=1715411947733" alt="biolxy">
          </span>
        
      </nav>
    
    <div class="md-post__meta md-meta">
      <ul class="md-meta__list">
        <li class="md-meta__item">
          <time datetime="2022-08-11 00:00:00+00:00">2022年8月11日星期四</time></li>
        
          <li class="md-meta__item">
            分类于
            
              <a href="%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6.html" class="md-meta__link">生物信息学</a></li>
        
        
          
          <li class="md-meta__item">
            
              需要 1 分钟阅读时间
            
          </li>
        
      </ul>
      
    </div>
  </header>
  <div class="md-post__content md-typeset">
    <h2 id="_1"><a class="toclink" href="../2022/08/11/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6%E4%B9%A0%E6%8A%80%E6%9C%AF%E8%B7%AF%E7%BA%BF%E7%94%9F%E4%BF%A1%E6%8C%87%E6%9C%88%E5%BD%95.html">生物信息学习技术路线(生信指月录)</a></h2>
<h3 id="1-linux"><a class="toclink" href="../2022/08/11/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6%E4%B9%A0%E6%8A%80%E6%9C%AF%E8%B7%AF%E7%BA%BF%E7%94%9F%E4%BF%A1%E6%8C%87%E6%9C%88%E5%BD%95.html#1-linux">1. Linux</a></h3>
<p>推荐 Linux就该这么学 （0-5章节，其他的可选择性学习）</p>
<ol>
<li>要求：<ol>
<li>掌握 shell 基础语法<ol>
<li>Linux基础命令</li>
<li>cd, ls, mkdir, pwd, time, df, cp, rm, sed, awk, wc, head, tail, more, history等</li>
</ol>
</li>
<li>目录如下：    <ol>
<li>第0章 咱们先来谈谈为什么要学习Linux系统</li>
<li>第1章 动手部署一台Linux操作系统</li>
<li>第2章 新手必须掌握的Linux命令</li>
<li>第3章 管道符、重定向与环境变量</li>
<li>第4章 Vim编辑器与Shell命令脚本</li>
<li>第5章 用户身份与文件权限</li>
</ol>
</li>
</ol>
</li>
</ol>
<h3 id="2-python"><a class="toclink" href="../2022/08/11/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6%E4%B9%A0%E6%8A%80%E6%9C%AF%E8%B7%AF%E7%BA%BF%E7%94%9F%E4%BF%A1%E6%8C%87%E6%9C%88%E5%BD%95.html#2-python">2. Python</a></h3>
<ol>
<li>掌握 基础语法，条件判断，数据类型，简单的数据结构</li>
<li>熟悉常用模块：sys, os, pandas, json, Bio, numpy, seaborn, pysam, argparse等</li>
<li>使用Python 读写文件（xls, xlsx, txt等文件）</li>
<li>了解 迭代器，生成器，装饰器等概念，并描述其适用的场景</li>
<li>python2 和 python3 中 range() 函数的区别</li>
<li>可以被next()函数调用并不断返回下一个值的对象称为迭代器：Iterator   (ref: https://www.liaoxuefeng.com/wiki/1016959663602400/1017323698112640)</li>
<li>在Python中，这种一边循环一边计算的机制，称为生成器：generator (ref: https://www.liaoxuefeng.com/wiki/1016959663602400/1017318207388128)
    熟悉 类 和 实例的概念，类函数，静态函数</li>
<li>可选：<ol>
<li>Python 多线程操作<ol>
<li>使用 pymysql 操作数据库</li>
<li>使用 python-docx等操作数据库</li>
<li>常见的设计模式及其使用场景</li>
<li>Python 数据结构与算法</li>
<li>陈斌老师 《数据结构与算法》B站课程</li>
<li>课件： 链接: https://pan.baidu.com/s/1srvCWOLsxEn3mOq1_TCmhQ 提取码: wtji</li>
</ol>
</li>
</ol>
</li>
</ol>
<h3 id="3"><a class="toclink" href="../2022/08/11/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6%E4%B9%A0%E6%8A%80%E6%9C%AF%E8%B7%AF%E7%BA%BF%E7%94%9F%E4%BF%A1%E6%8C%87%E6%9C%88%E5%BD%95.html#3">3. 生物信息学课程</a></h3>
<ol>
<li>了解相关概念, 专有名词</li>
<li>了解相关概念背后的 生物信息算法，统计原理</li>
<li>【 山东大学】生物信息学 B站课程</li>
<li>【北京大学】生物信息学：导论与方法 B站课程</li>
<li>统计学
    ⅰ. https://www.yuque.com/biolxy/bioinfo/zgl1ml</li>
</ol>
<h3 id="4-biostar-httpswwwbiostarhandbookcom"><a class="toclink" href="../2022/08/11/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6%E4%B9%A0%E6%8A%80%E6%9C%AF%E8%B7%AF%E7%BA%BF%E7%94%9F%E4%BF%A1%E6%8C%87%E6%9C%88%E5%BD%95.html#4-biostar-httpswwwbiostarhandbookcom">4. BioStar 实操练习 https://www.biostarhandbook.com/</a></h3>
<ul>
<li>可以看生信媛翻译的 BIOSTAR课程</li>
</ul>
<h3 id="5"><a class="toclink" href="../2022/08/11/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6%E4%B9%A0%E6%8A%80%E6%9C%AF%E8%B7%AF%E7%BA%BF%E7%94%9F%E4%BF%A1%E6%8C%87%E6%9C%88%E5%BD%95.html#5">5. 肿瘤外显子数据分析指南</a></h3>
<ul>
<li>https://www.yuque.com/biotrainee/wes</li>
</ul>
<p><img alt="sad" src="https://www.biostarhandbook.com/images/what-is-a-biostar.png" /></p>
    
  </div>
</article>
      
        <article class="md-post md-post--excerpt">
  <header class="md-post__header">
    
      <nav class="md-post__authors md-typeset">
        
          <span class="md-author">
            <img src="https://gravatar.com/avatar/d63d3425433de59d2c3e901977a581c7?size=256&cache=1715411947733" alt="biolxy">
          </span>
        
      </nav>
    
    <div class="md-post__meta md-meta">
      <ul class="md-meta__list">
        <li class="md-meta__item">
          <time datetime="2021-08-22 00:00:00+00:00">2021年8月22日星期日</time></li>
        
          <li class="md-meta__item">
            分类于
            
              <a href="%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6.html" class="md-meta__link">生物信息学</a></li>
        
        
          
          <li class="md-meta__item">
            
              需要 2 分钟阅读时间
            
          </li>
        
      </ul>
      
    </div>
  </header>
  <div class="md-post__content md-typeset">
    <h2 id="_1"><a class="toclink" href="../2021/08/22/%E8%8E%B7%E5%8F%96%E5%90%AF%E5%8A%A8%E5%AD%90%E5%BA%8F%E5%88%97.html">获取启动子序列</a></h2>
<h4 id="_2"><a class="toclink" href="../2021/08/22/%E8%8E%B7%E5%8F%96%E5%90%AF%E5%8A%A8%E5%AD%90%E5%BA%8F%E5%88%97.html#_2">获取启动子序列</a></h4>
<ol>
<li>获取物种基因组序列genome.fa文件</li>
<li>获取物种注释信息gff文件</li>
</ol>
<p>输入参数：
positionFile, genomeFa, outfa
（pos坐标文件，包含四列 chrom start end strand）</p>
<p>使用Python脚本
<div class="language-python highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">Python</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span></pre></div></td><td class="code"><div><pre><span></span><code><span id="__span-2-1"><span class="ch">#!/usr/bin/env python</span>
</span><span id="__span-2-2"><span class="c1"># -*- encoding: utf-8 -*-</span>
</span><span id="__span-2-3"><span class="sd">&#39;&#39;&#39;</span>
</span><span id="__span-2-4"><span class="sd">@File    :   Promoter.py</span>
</span><span id="__span-2-5"><span class="sd">@Time    :   2021/08/15 20:16:07</span>
</span><span id="__span-2-6"><span class="sd">@Author  :   biolxy </span>
</span><span id="__span-2-7"><span class="sd">@Version :   1.0</span>
</span><span id="__span-2-8"><span class="sd">@Contact :   biolxy@aliyun.com</span>
</span><span id="__span-2-9"><span class="sd">@Desc    :   None</span>
</span><span id="__span-2-10"><span class="sd">@Usage   :   python Promoter.py pos.Lee.txt /home/lixy/soybean/other/glyma.Lee.gnm1.BXNC.genome_main.fna out.Lee.fa</span>
</span><span id="__span-2-11"><span class="sd">&#39;&#39;&#39;</span>
</span><span id="__span-2-12"><span class="kn">from</span><span class="w"> </span><span class="nn">Bio.Seq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Seq</span>
</span><span id="__span-2-13"><span class="kn">from</span><span class="w"> </span><span class="nn">Bio.SeqRecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">SeqRecord</span>
</span><span id="__span-2-14"><span class="kn">from</span><span class="w"> </span><span class="nn">Bio</span><span class="w"> </span><span class="kn">import</span> <span class="n">SeqIO</span>
</span><span id="__span-2-15"><span class="kn">import</span><span class="w"> </span><span class="nn">sys</span>
</span><span id="__span-2-16">
</span><span id="__span-2-17">
</span><span id="__span-2-18"><span class="k">class</span><span class="w"> </span><span class="nc">Promoter</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
</span><span id="__span-2-19">    <span class="nd">@staticmethod</span>
</span><span id="__span-2-20">    <span class="k">def</span><span class="w"> </span><span class="nf">get_promoter_by_genome</span><span class="p">(</span><span class="n">positionFile</span><span class="p">,</span> <span class="n">genomeFa</span><span class="p">,</span> <span class="n">outfa</span><span class="p">,</span> <span class="n">length</span><span class="o">=</span><span class="mi">2000</span><span class="p">):</span>
</span><span id="__span-2-21"><span class="w">        </span><span class="sd">&quot;&quot;&quot;positionFile</span>
</span><span id="__span-2-22"><span class="sd">        Chr20   35557820    35562522    +</span>
</span><span id="__span-2-23">
</span><span id="__span-2-24"><span class="sd">        length: 2000</span>
</span><span id="__span-2-25"><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="__span-2-26">        <span class="n">dict1</span> <span class="o">=</span> <span class="p">{}</span>
</span><span id="__span-2-27">        <span class="n">pos_list</span> <span class="o">=</span> <span class="n">Promoter</span><span class="o">.</span><span class="n">get_pos_list</span><span class="p">(</span><span class="n">positionFile</span><span class="p">)</span>
</span><span id="__span-2-28">        <span class="k">for</span> <span class="n">seq_record</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">genomeFa</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">):</span>
</span><span id="__span-2-29">            <span class="n">seqid</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">seq_record</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
</span><span id="__span-2-30">            <span class="k">for</span> <span class="n">tmp_l</span> <span class="ow">in</span> <span class="n">pos_list</span><span class="p">:</span>
</span><span id="__span-2-31">                <span class="k">if</span> <span class="n">seqid</span> <span class="o">==</span> <span class="n">tmp_l</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span>
</span><span id="__span-2-32">                    <span class="n">seq</span> <span class="o">=</span> <span class="n">seq_record</span><span class="o">.</span><span class="n">seq</span>
</span><span id="__span-2-33">                    <span class="n">s</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">tmp_l</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="o">-</span> <span class="n">length</span>
</span><span id="__span-2-34">                    <span class="n">e</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">tmp_l</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
</span><span id="__span-2-35">
</span><span id="__span-2-36">                    <span class="k">if</span> <span class="s1">&#39;-&#39;</span> <span class="o">==</span> <span class="n">tmp_l</span><span class="p">[</span><span class="mi">3</span><span class="p">]:</span>
</span><span id="__span-2-37">                        <span class="n">s</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">tmp_l</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
</span><span id="__span-2-38">                        <span class="n">e</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">tmp_l</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span> <span class="o">+</span> <span class="n">length</span>
</span><span id="__span-2-39">                    <span class="c1"># print(s, e)</span>
</span><span id="__span-2-40">                    <span class="n">seq</span> <span class="o">=</span> <span class="n">seq</span><span class="p">[</span><span class="n">s</span><span class="p">:</span><span class="n">e</span><span class="p">]</span>
</span><span id="__span-2-41">                    <span class="k">if</span> <span class="s1">&#39;-&#39;</span> <span class="o">==</span> <span class="n">tmp_l</span><span class="p">[</span><span class="mi">3</span><span class="p">]:</span>
</span><span id="__span-2-42">                        <span class="n">seq</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span> <span class="c1"># 反向互补</span>
</span><span id="__span-2-43">                    <span class="n">n_seqid</span> <span class="o">=</span> <span class="n">seqid</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">tmp_l</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="o">+</span> <span class="s2">&quot;:&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">tmp_l</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">tmp_l</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
</span><span id="__span-2-44">                    <span class="n">dict1</span><span class="p">[</span><span class="n">n_seqid</span><span class="p">]</span> <span class="o">=</span> <span class="n">seq</span>
</span><span id="__span-2-45">
</span><span id="__span-2-46">        <span class="n">records</span> <span class="o">=</span> <span class="p">[]</span>
</span><span id="__span-2-47">        <span class="k">for</span> <span class="n">seqid</span> <span class="ow">in</span> <span class="n">dict1</span><span class="p">:</span>
</span><span id="__span-2-48">            <span class="n">seq</span> <span class="o">=</span> <span class="n">dict1</span><span class="p">[</span><span class="n">seqid</span><span class="p">]</span>
</span><span id="__span-2-49">            <span class="n">rec1</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="n">seqid</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
</span><span id="__span-2-50">            <span class="n">records</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">rec1</span><span class="p">)</span>
</span><span id="__span-2-51">        <span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">records</span><span class="p">,</span> <span class="n">outfa</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
</span><span id="__span-2-52">
</span><span id="__span-2-53">
</span><span id="__span-2-54">    <span class="nd">@staticmethod</span>
</span><span id="__span-2-55">    <span class="k">def</span><span class="w"> </span><span class="nf">get_pos_list</span><span class="p">(</span><span class="n">infile</span><span class="p">):</span>
</span><span id="__span-2-56">        <span class="n">list_</span> <span class="o">=</span> <span class="p">[]</span>
</span><span id="__span-2-57">        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">infile</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">ff</span><span class="p">:</span>
</span><span id="__span-2-58">            <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">ff</span><span class="p">:</span>
</span><span id="__span-2-59">                <span class="n">line</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
</span><span id="__span-2-60">                <span class="n">tag</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span> <span class="c1"># chrom start end strand</span>
</span><span id="__span-2-61">                <span class="n">list_</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tag</span><span class="p">)</span>
</span><span id="__span-2-62">        <span class="k">return</span> <span class="n">list_</span>
</span><span id="__span-2-63">
</span><span id="__span-2-64">
</span><span id="__span-2-65"><span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s2">&quot;__main__&quot;</span><span class="p">:</span>
</span><span id="__span-2-66">    <span class="n">positionFile</span><span class="p">,</span> <span class="n">genomeFa</span><span class="p">,</span> <span class="n">outfa</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>
</span><span id="__span-2-67">    <span class="n">Promoter</span><span class="o">.</span><span class="n">get_promoter_by_genome</span><span class="p">(</span><span class="n">positionFile</span><span class="p">,</span> <span class="n">genomeFa</span><span class="p">,</span> <span class="n">outfa</span><span class="p">)</span>
</span></code></pre></div></td></tr></table></div></p>
    
  </div>
</article>
      
        <article class="md-post md-post--excerpt">
  <header class="md-post__header">
    
      <nav class="md-post__authors md-typeset">
        
          <span class="md-author">
            <img src="https://gravatar.com/avatar/d63d3425433de59d2c3e901977a581c7?size=256&cache=1715411947733" alt="biolxy">
          </span>
        
      </nav>
    
    <div class="md-post__meta md-meta">
      <ul class="md-meta__list">
        <li class="md-meta__item">
          <time datetime="2020-10-28 00:00:00+00:00">2020年10月28日星期三</time></li>
        
          <li class="md-meta__item">
            分类于
            
              <a href="%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6.html" class="md-meta__link">生物信息学</a></li>
        
        
          
          <li class="md-meta__item">
            
              需要 3 分钟阅读时间
            
          </li>
        
      </ul>
      
    </div>
  </header>
  <div class="md-post__content md-typeset">
    <h2 id="reads-count-tpm-fpkm"><a class="toclink" href="../2020/10/28/%E7%94%B1-reads-count-%E6%95%B0%E8%AE%A1%E7%AE%97-tpm-%E5%92%8Cfpkm.html">由 reads count 数计算 TPM 和FPKM</a></h2>
<h5 id="reads-count-tpm-fpkm_1"><a class="toclink" href="../2020/10/28/%E7%94%B1-reads-count-%E6%95%B0%E8%AE%A1%E7%AE%97-tpm-%E5%92%8Cfpkm.html#reads-count-tpm-fpkm_1">由 reads count 数计算 TPM 和FPKM</a></h5>
<ul>
<li>下载gff文件</li>
<li>http://venanciogroup.uenf.br/cgi-bin/gmax_atlas/download_by_pub.cgi 去这里下载reads count 数据</li>
</ul>
<div class="language-r highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">S</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span></pre></div></td><td class="code"><div><pre><span></span><code><span id="__span-3-1"><span class="nf">library</span><span class="p">(</span><span class="s">&quot;GenomicFeatures&quot;</span><span class="p">)</span>
</span><span id="__span-3-2"><span class="nf">library</span><span class="p">(</span><span class="s">&quot;dplyr&quot;</span><span class="p">)</span>
</span><span id="__span-3-3"><span class="n">path1</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">setwd</span><span class="p">(</span><span class="s">&quot;C:/Users/95656/Desktop/hyp/TZX/&quot;</span><span class="p">)</span>
</span><span id="__span-3-4"><span class="nf">setwd</span><span class="p">(</span><span class="n">path1</span><span class="p">)</span>
</span><span id="__span-3-5"><span class="c1"># 计算基因length</span>
</span><span id="__span-3-6">
</span><span id="__span-3-7"><span class="n">gffpath</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Gmax_275_Wm82.a2.v1.gene.gff3&quot;</span>
</span><span id="__span-3-8"><span class="nf">print</span><span class="p">(</span><span class="n">gffpath</span><span class="p">)</span>
</span><span id="__span-3-9"><span class="n">txdb</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">makeTxDbFromGFF</span><span class="p">(</span><span class="n">gffpath</span><span class="p">,</span><span class="n">format</span><span class="o">=</span><span class="s">&quot;gff&quot;</span><span class="p">)</span>
</span><span id="__span-3-10"><span class="n">exons_gene</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">exonsBy</span><span class="p">(</span><span class="n">txdb</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;gene&quot;</span><span class="p">)</span>
</span><span id="__span-3-11"><span class="n">exons_gene_lens</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">lapply</span><span class="p">(</span><span class="n">exons_gene</span><span class="p">,</span><span class="kr">function</span><span class="p">(</span><span class="n">x</span><span class="p">){</span><span class="nf">sum</span><span class="p">(</span><span class="nf">width</span><span class="p">(</span><span class="nf">reduce</span><span class="p">(</span><span class="n">x</span><span class="p">)))})</span>
</span><span id="__span-3-12"><span class="n">length1</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">t</span><span class="p">(</span><span class="nf">as.data.frame</span><span class="p">(</span><span class="n">exons_gene_lens</span><span class="p">))</span>
</span><span id="__span-3-13"><span class="nf">write.table</span><span class="p">(</span><span class="n">length1</span><span class="p">,</span><span class="s">&quot;C:/Users/95656/Desktop/hyp/TZX/length.txt&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;\t&quot;</span><span class="p">)</span>
</span><span id="__span-3-14">
</span><span id="__span-3-15"><span class="c1"># 合并 reads count 数据和length</span>
</span><span id="__span-3-16"><span class="n">length1</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">read.table</span><span class="p">(</span><span class="s">&quot;C:/Users/95656/Desktop/hyp/TZX/length.txt&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">header</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="bp">T</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">)</span>
</span><span id="__span-3-17"><span class="n">countsfile</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;C:/Users/95656/Desktop/hyp/TZX/PRJNA238493_BP_raw_count.expression.tsv&quot;</span>
</span><span id="__span-3-18"><span class="n">count1</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">read.table</span><span class="p">(</span><span class="n">countsfile</span><span class="p">,</span><span class="w"> </span><span class="n">header</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="bp">T</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;\t&quot;</span><span class="p">)</span>
</span><span id="__span-3-19"><span class="n">a</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">inner_join</span><span class="p">(</span><span class="n">count1</span><span class="p">,</span><span class="w"> </span><span class="n">length1</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="o">=</span><span class="s">&quot;Locus_Name&quot;</span><span class="p">)</span>
</span><span id="__span-3-20"><span class="n">a</span><span class="p">[</span><span class="nf">is.na</span><span class="p">(</span><span class="n">a</span><span class="p">)]</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="w"> </span><span class="c1"># 设置 NA值为0</span>
</span><span id="__span-3-21"><span class="n">a</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">a</span><span class="p">[,</span><span class="w"> </span><span class="m">-28</span><span class="p">]</span><span class="w">  </span><span class="c1"># 删除第28列  Locus_Name.1</span>
</span><span id="__span-3-22"><span class="nf">write.table</span><span class="p">(</span><span class="n">a</span><span class="p">,</span><span class="s">&quot;C:/Users/95656/Desktop/hyp/TZX/countsdata.txt&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;\t&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">row.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">)</span>
</span><span id="__span-3-23">
</span><span id="__span-3-24"><span class="c1"># 按照对应关系给countdata columns 改名，改好的文件保存为 re_countsdata.txt</span>
</span><span id="__span-3-25"><span class="c1"># {&#39;SAMN02644800&#39;: &#39;root&#39;, &#39;SAMN02644811&#39;: &#39;leaf3&#39;, &#39;SAMN02644826&#39;: &#39;seed4&#39;, &#39;SAMN02644825&#39;: &#39;seed3&#39;, &#39;SAMN02644822&#39;: &#39;podseed3&#39;, &#39;SAMN02644821&#39;: &#39;podseed2&#39;, &#39;SAMN02644820&#39;: &#39;podseed1&#39;, &#39;SAMN02644803&#39;: &#39;stem1&#39;, &#39;SAMN02644806&#39;: &#39;leafbud1&#39;, &#39;SAMN02644804&#39;: &#39;stem2&#39;, &#39;SAMN02644813&#39;: &#39;flower2&#39;, &#39;SAMN02644814&#39;: &#39;flower3&#39;, &#39;SAMN02644816&#39;: &#39;flower5&#39;, &#39;SAMN02644815&#39;: &#39;flower4&#39;, &#39;SAMN02644817&#39;: &#39;pod1&#39;, &#39;SAMN02644823&#39;: &#39;seed1&#39;, &#39;SAMN02644818&#39;: &#39;pod2&#39;, &#39;SAMN02644819&#39;: &#39;pod3&#39;, &#39;SAMN02644824&#39;: &#39;seed2&#39;, &#39;SAMN02644809&#39;: &#39;leaf1&#39;, &#39;SAMN02644807&#39;: &#39;leafbud2&#39;, &#39;SAMN02644802&#39;: &#39;cotyledon2&#39;, &#39;SAMN02644810&#39;: &#39;leaf2&#39;, &#39;SAMN02644805&#39;: &#39;shoot meristem&#39;, &#39;SAMN02644812&#39;: &#39;flower1&#39;, &#39;SAMN02644827&#39;: &#39;seed5&#39;}</span>
</span><span id="__span-3-26">
</span><span id="__span-3-27"><span class="c1"># 计算 TPM 和 FPKM</span>
</span><span id="__span-3-28">
</span><span id="__span-3-29"><span class="nf">setwd</span><span class="p">(</span><span class="s">&quot;C:/Users/95656/Desktop/hyp/TZX/&quot;</span><span class="p">)</span>
</span><span id="__span-3-30"><span class="n">countfile</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;C:/Users/95656/Desktop/hyp/TZX/re_countsdata.txt&quot;</span>
</span><span id="__span-3-31"><span class="n">mycounts</span><span class="o">&lt;-</span><span class="nf">read.table</span><span class="p">(</span><span class="n">countfile</span><span class="p">,</span><span class="w"> </span><span class="n">header</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="bp">T</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">)</span>
</span><span id="__span-3-32"><span class="nf">head</span><span class="p">(</span><span class="n">mycounts</span><span class="p">)</span>
</span><span id="__span-3-33"><span class="nf">rownames</span><span class="p">(</span><span class="n">mycounts</span><span class="p">)</span><span class="o">&lt;-</span><span class="n">mycounts</span><span class="p">[,</span><span class="m">1</span><span class="p">]</span>
</span><span id="__span-3-34"><span class="n">mycounts</span><span class="o">&lt;-</span><span class="n">mycounts</span><span class="p">[,</span><span class="m">-1</span><span class="p">]</span>
</span><span id="__span-3-35"><span class="nf">head</span><span class="p">(</span><span class="n">mycounts</span><span class="p">)</span>
</span><span id="__span-3-36"><span class="c1"># TPM</span>
</span><span id="__span-3-37"><span class="n">kb</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">mycounts</span><span class="o">$</span><span class="n">length</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="m">1000</span>
</span><span id="__span-3-38"><span class="n">countdata</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">mycounts</span><span class="p">[,</span><span class="m">1</span><span class="o">:</span><span class="m">26</span><span class="p">]</span>
</span><span id="__span-3-39"><span class="n">rpk</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">countdata</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="n">kb</span>
</span><span id="__span-3-40"><span class="n">tpm</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">t</span><span class="p">(</span><span class="nf">t</span><span class="p">(</span><span class="n">rpk</span><span class="p">)</span><span class="o">/</span><span class="nf">colSums</span><span class="p">(</span><span class="n">rpk</span><span class="p">)</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="m">1000000</span><span class="p">)</span>
</span><span id="__span-3-41"><span class="nf">head</span><span class="p">(</span><span class="n">tpm</span><span class="p">)</span>
</span><span id="__span-3-42"><span class="nf">write.table</span><span class="p">(</span><span class="n">tpm</span><span class="p">,</span><span class="n">file</span><span class="o">=</span><span class="s">&quot;my_TPM.txt&quot;</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">,</span><span class="n">quote</span><span class="o">=</span><span class="bp">F</span><span class="p">)</span>
</span><span id="__span-3-43"><span class="c1"># FPKM</span>
</span><span id="__span-3-44"><span class="n">fpkm</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">t</span><span class="p">(</span><span class="nf">t</span><span class="p">(</span><span class="n">rpk</span><span class="p">)</span><span class="o">/</span><span class="nf">colSums</span><span class="p">(</span><span class="n">countdata</span><span class="p">)</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="m">10</span><span class="o">^</span><span class="m">6</span><span class="p">)</span>
</span><span id="__span-3-45"><span class="nf">head</span><span class="p">(</span><span class="n">fpkm</span><span class="p">)</span>
</span><span id="__span-3-46"><span class="nf">write.table</span><span class="p">(</span><span class="n">fpkm</span><span class="p">,</span><span class="n">file</span><span class="o">=</span><span class="s">&quot;my_FPKM.txt&quot;</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">,</span><span class="n">quote</span><span class="o">=</span><span class="bp">F</span><span class="p">)</span>
</span><span id="__span-3-47">
</span><span id="__span-3-48">
</span><span id="__span-3-49"><span class="c1">## FPKM转化为TPM</span>
</span><span id="__span-3-50"><span class="c1"># fpkm_to_tpm = t(t(fpkm)/colSums(fpkm))*10^6</span>
</span><span id="__span-3-51"><span class="c1"># head(fpkm_to_tpm)</span>
</span></code></pre></div></td></tr></table></div>
<h6 id="_1"><a class="toclink" href="../2020/10/28/%E7%94%B1-reads-count-%E6%95%B0%E8%AE%A1%E7%AE%97-tpm-%E5%92%8Cfpkm.html#_1">计算 共表达相关性</a></h6>
<div class="language-text highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">Text Only</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span></pre></div></td><td class="code"><div><pre><span></span><code><span id="__span-4-1">path1 = &quot;C:\\Users\\95656\\Desktop\\hyp\\比较&quot;
</span><span id="__span-4-2">setwd(path1)
</span><span id="__span-4-3">rowdata &lt;- read.table(choose.files(),header=T,sep=&quot;\t&quot;)
</span><span id="__span-4-4"># exp1 = rowdata[,5:]
</span><span id="__span-4-5"># &quot;cotyledon1&quot;, &quot;leafbud3&quot;,
</span><span id="__span-4-6">exp1 = rowdata[c(&quot;cotyledon2&quot;, &quot;flower1&quot;, &quot;flower2&quot;, &quot;flower3&quot;, &quot;flower4&quot;, &quot;flower5&quot;, &quot;leaf1&quot;, &quot;leaf2&quot;, &quot;leaf3&quot;, &quot;leafbud1&quot;, &quot;leafbud2&quot;, &quot;pod1&quot;, &quot;pod2&quot;, &quot;pod3&quot;, &quot;podseed1&quot;, &quot;podseed2&quot;, &quot;podseed3&quot;, &quot;root&quot;, &quot;seed1&quot;, &quot;seed2&quot;, &quot;seed3&quot;, &quot;seed4&quot;, &quot;seed5&quot;, &quot;shootmeristem&quot;, &quot;stem1&quot;, &quot;stem2&quot;)]
</span><span id="__span-4-7">rownames(exp1)&lt;-rowdata$Name
</span><span id="__span-4-8">exp1[is.na(exp1)] = 0 # 设置 NA值为0
</span><span id="__span-4-9">exp1 = as.matrix(exp1)
</span><span id="__span-4-10">geneid &lt;- rowdata[,&#39;Name&#39;]
</span><span id="__span-4-11">TCP = geneid[1:5] 
</span><span id="__span-4-12">Gma = geneid[1:5]
</span><span id="__span-4-13">outfile = choose.files()
</span><span id="__span-4-14">for(i in 1:length(TCP))
</span><span id="__span-4-15">{
</span><span id="__span-4-16">    for(j in 1:length(Gma))
</span><span id="__span-4-17">    {
</span><span id="__span-4-18">        x &lt;- as.numeric(exp1[i,])
</span><span id="__span-4-19">        y &lt;- as.numeric(exp1[j,])
</span><span id="__span-4-20">        if(sum(exp1[i,])*sum(exp1[j,]) != 0)#取出表达值都为0的行，如果要严格一点，想要把表达值出现0的行都去掉的话可以把求和函数sum()换成求积函数prod(x)
</span><span id="__span-4-21">        {
</span><span id="__span-4-22">            gene_cor &lt;- cor.test(x,y,alternative = &quot;two.sided&quot;,method = &quot;pearson&quot;)
</span><span id="__span-4-23">            gene_p &lt;- gene_cor$p.value
</span><span id="__span-4-24">            gene_e &lt;- gene_cor$estimate
</span><span id="__span-4-25">            if(abs(gene_e) &gt;= 0 &amp;&amp; gene_p &lt;= 1)
</span><span id="__span-4-26">            {
</span><span id="__span-4-27">                res&lt;-paste(TCP[i],Gma[j],gene_e,gene_p,sep=&quot;\t&quot;)
</span><span id="__span-4-28">                write.table(res, outfile,sep=&quot;\t&quot;,col.names=F,row.names=F,quote=F,append=T)
</span><span id="__span-4-29">            }
</span><span id="__span-4-30">        }else{
</span><span id="__span-4-31">            print(&quot;something is error&quot;)
</span><span id="__span-4-32">        }
</span><span id="__span-4-33">    }
</span><span id="__span-4-34">}
</span></code></pre></div></td></tr></table></div>
<h6 id="python-df"><a class="toclink" href="../2020/10/28/%E7%94%B1-reads-count-%E6%95%B0%E8%AE%A1%E7%AE%97-tpm-%E5%92%8Cfpkm.html#python-df">用 python df 计算相关性</a></h6>
<div class="language-python highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">Python</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span></pre></div></td><td class="code"><div><pre><span></span><code><span id="__span-5-1"><span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
</span><span id="__span-5-2"><span class="kn">from</span><span class="w"> </span><span class="nn">scipy</span><span class="w"> </span><span class="kn">import</span> <span class="n">stats</span>
</span><span id="__span-5-3"><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
</span><span id="__span-5-4">
</span><span id="__span-5-5">
</span><span id="__span-5-6"><span class="k">def</span><span class="w"> </span><span class="nf">get_dfcorr</span><span class="p">(</span><span class="n">df1</span><span class="p">,</span> <span class="n">df2</span><span class="p">):</span>
</span><span id="__span-5-7"><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-5-8"><span class="sd">    按行计算df1(gene1) 和 df2(gene2) 中数据的相关性, 其中 df1、2的列数可以不相等，method=&quot;pearson&quot;</span>
</span><span id="__span-5-9"><span class="sd">    100W 1m32s</span>
</span><span id="__span-5-10"><span class="sd">    df 中 列为组织名称，行为基因名称</span>
</span><span id="__span-5-11"><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="__span-5-12">    <span class="n">df1</span> <span class="o">=</span> <span class="n">df1</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="o">~</span><span class="p">(</span><span class="n">df1</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span> <span class="p">:]</span>
</span><span id="__span-5-13">    <span class="n">df2</span> <span class="o">=</span> <span class="n">df2</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="o">~</span><span class="p">(</span><span class="n">df2</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span> <span class="p">:]</span>
</span><span id="__span-5-14">    <span class="n">df1index</span> <span class="o">=</span> <span class="n">df1</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
</span><span id="__span-5-15">    <span class="n">df2index</span> <span class="o">=</span> <span class="n">df2</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
</span><span id="__span-5-16">    <span class="n">list1</span><span class="p">,</span> <span class="n">list2</span> <span class="o">=</span> <span class="n">df1</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">(),</span> <span class="n">df2</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
</span><span id="__span-5-17">    <span class="n">list_r</span> <span class="o">=</span> <span class="p">[]</span>
</span><span id="__span-5-18">    <span class="k">for</span> <span class="n">n</span><span class="p">,</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">list1</span><span class="p">):</span>
</span><span id="__span-5-19">        <span class="k">for</span> <span class="n">m</span><span class="p">,</span> <span class="n">y</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">list2</span><span class="p">):</span>
</span><span id="__span-5-20">            <span class="n">a</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">float</span><span class="p">)</span>
</span><span id="__span-5-21">            <span class="n">b</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">y</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">float</span><span class="p">)</span>
</span><span id="__span-5-22">            <span class="n">r</span><span class="p">,</span> <span class="n">p</span> <span class="o">=</span> <span class="n">stats</span><span class="o">.</span><span class="n">pearsonr</span><span class="p">(</span><span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">)</span>
</span><span id="__span-5-23">            <span class="n">listtmp</span> <span class="o">=</span> <span class="p">[</span><span class="n">df1index</span><span class="p">[</span><span class="n">n</span><span class="p">],</span> <span class="n">df2index</span><span class="p">[</span><span class="n">m</span><span class="p">],</span> <span class="n">r</span><span class="p">,</span> <span class="n">p</span><span class="p">]</span>
</span><span id="__span-5-24">            <span class="n">list_r</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">listtmp</span><span class="p">)</span>
</span><span id="__span-5-25">    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">list_r</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;gene1&quot;</span><span class="p">,</span> <span class="s2">&quot;gene2&quot;</span><span class="p">,</span> <span class="s2">&quot;correlation&quot;</span><span class="p">,</span> <span class="s2">&quot;pvalue&quot;</span><span class="p">])</span>
</span><span id="__span-5-26">    <span class="k">return</span> <span class="n">df</span>
</span></code></pre></div></td></tr></table></div>
    
  </div>
</article>
      
        <article class="md-post md-post--excerpt">
  <header class="md-post__header">
    
      <nav class="md-post__authors md-typeset">
        
          <span class="md-author">
            <img src="https://gravatar.com/avatar/d63d3425433de59d2c3e901977a581c7?size=256&cache=1715411947733" alt="biolxy">
          </span>
        
      </nav>
    
    <div class="md-post__meta md-meta">
      <ul class="md-meta__list">
        <li class="md-meta__item">
          <time datetime="2018-08-01 00:00:00+00:00">2018年8月1日星期三</time></li>
        
          <li class="md-meta__item">
            分类于
            
              <a href="%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6.html" class="md-meta__link">生物信息学</a></li>
        
        
          
          <li class="md-meta__item">
            
              需要 5 分钟阅读时间
            
          </li>
        
      </ul>
      
    </div>
  </header>
  <div class="md-post__content md-typeset">
    <h2 id="linux"><a class="toclink" href="../2018/08/01/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6%E5%B8%B8%E7%94%A8%E7%9A%84linux%E5%91%BD%E4%BB%A4.html">生物信息学常用的Linux命令</a></h2>
<h4 id="1"><a class="toclink" href="../2018/08/01/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6%E5%B8%B8%E7%94%A8%E7%9A%84linux%E5%91%BD%E4%BB%A4.html#1">1、 前言</a></h4>
<p>东西有点杂乱，还需要慢慢整理。其中的 sed awk perl 等命令我觉得都可以单独分出来讲了</p>
<h4 id="2-blast"><a class="toclink" href="../2018/08/01/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6%E5%B8%B8%E7%94%A8%E7%9A%84linux%E5%91%BD%E4%BB%A4.html#2-blast">2、 构建blast数据库</a></h4>
<div class="language-shell highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">Bash</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span></pre></div></td><td class="code"><div><pre><span></span><code><span id="__span-0-1">makeblastdb<span class="w"> </span>-in<span class="w"> </span>../ATaa.fa<span class="w"> </span>-dbtype<span class="w"> </span>prot<span class="w"> </span>-title<span class="w"> </span>ATaa-parse_seqids<span class="w"> </span>-out<span class="w"> </span>ATdb<span class="w">   </span><span class="c1">#可用的蛋白库</span>
</span><span id="__span-0-2">makeblastdb<span class="w"> </span>-in<span class="w"> </span>../ATaa.fa<span class="w"> </span>-dbtype<span class="w"> </span>nucl<span class="w"> </span>-title<span class="w"> </span>ATaa-parse_seqids<span class="w"> </span>-out<span class="w"> </span>ATdb<span class="w">  </span><span class="c1">#可用的核酸白库</span>
</span><span id="__span-0-3">blastp<span class="w"> </span>-db<span class="w"> </span>/home/hanyapeng/prot/DB/Mtrdb<span class="w"> </span>-query<span class="w"> </span>At.fa<span class="w"> </span>-outfmt<span class="w"> </span><span class="m">6</span><span class="w"> </span>-evalue<span class="w"> </span>1e-20<span class="w"> </span>-out<span class="w"> </span>Mtr.blast<span class="w"> </span>-num_threads<span class="w"> </span><span class="m">8</span>
</span><span id="__span-0-4">blastn<span class="w"> </span>-query<span class="w"> </span>seq.fasta<span class="w"> </span>-out<span class="w"> </span>seq.blast<span class="w"> </span>-db<span class="w"> </span>dbname<span class="w"> </span>-outfmt<span class="w"> </span><span class="m">6</span><span class="w"> </span>-evalue<span class="w"> </span>1e-5<span class="w"> </span>-num_descriptions<span class="w"> </span><span class="m">10</span><span class="w"> </span>-num_threads<span class="w"> </span><span class="m">8</span>
</span></code></pre></div></td></tr></table></div>
<h4 id="3"><a class="toclink" href="../2018/08/01/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6%E5%B8%B8%E7%94%A8%E7%9A%84linux%E5%91%BD%E4%BB%A4.html#3">3、 碱基/蛋白序列比对</a></h4>
<div class="language-text highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">Text Only</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span></pre></div></td><td class="code"><div><pre><span></span><code><span id="__span-1-1">muscle -in AtLj.fa -out AtLj.phy  # -out，输出默认的fasta格式
</span><span id="__span-1-2">muscle -in At.fasta -phyiout At.phy  # -phyiout，则输出sequential phylip格式
</span><span id="__span-1-3">mafft aa.fa &gt; aa.fas
</span></code></pre></div></td></tr></table></div>
<h4 id="4"><a class="toclink" href="../2018/08/01/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6%E5%B8%B8%E7%94%A8%E7%9A%84linux%E5%91%BD%E4%BB%A4.html#4">4、 画树</a></h4>
<div class="language-shell highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">Bash</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span></pre></div></td><td class="code"><div><pre><span></span><code><span id="__span-2-1">nohup<span class="w"> </span>raxmlHPC-PTHREADS<span class="w"> </span>-s<span class="w"> </span>ALO.phy<span class="w"> </span>-n<span class="w"> </span>ALO<span class="w"> </span>-m<span class="w"> </span>PROTCATJTT<span class="w"> </span>-f<span class="w"> </span>a<span class="w"> </span>-N<span class="w"> </span><span class="m">100</span><span class="w"> </span>-p<span class="w"> </span><span class="m">12345</span><span class="w"> </span>-x<span class="w"> </span><span class="m">1234567</span><span class="w"> </span>-T<span class="w"> </span><span class="m">8</span><span class="w"> </span><span class="p">&amp;</span><span class="w">  </span><span class="c1"># RaxML树</span>
</span><span id="__span-2-2">nohup<span class="w"> </span>../software/PhyML_3.0_linux64<span class="w"> </span>-i<span class="w"> </span>AGO.phy<span class="w"> </span>-d<span class="w"> </span>aa<span class="w"> </span>-b<span class="w"> </span><span class="m">0</span><span class="w">  </span>-f<span class="w"> </span>e<span class="w"> </span>-s<span class="w"> </span>BEST<span class="w"> </span>-c<span class="w"> </span><span class="m">4</span><span class="w"> </span>-m<span class="w"> </span>VT<span class="w"> </span>-v<span class="w"> </span>e<span class="w"> </span>-a<span class="w"> </span>e<span class="w"> </span>-o<span class="w"> </span>lr<span class="w"> </span>--no_memory_check<span class="w"> </span><span class="p">&amp;</span><span class="w">  </span><span class="c1"># Phyml 树</span>
</span><span id="__span-2-3">../software/FastTree<span class="w"> </span>TCP123.phy<span class="w">  </span>&gt;<span class="w"> </span>TCP.tree<span class="w">   </span><span class="c1">#fasttree 画树</span>
</span></code></pre></div></td></tr></table></div>
<h4 id="5-cdspep"><a class="toclink" href="../2018/08/01/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6%E5%B8%B8%E7%94%A8%E7%9A%84linux%E5%91%BD%E4%BB%A4.html#5-cdspep">5、 提取CDS/pep序列</a></h4>
<div class="language-shell highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">Bash</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span></pre></div></td><td class="code"><div><pre><span></span><code><span id="__span-3-1">nohup<span class="w"> </span>hmmsearch<span class="w"> </span>-E<span class="w"> </span>1e-5<span class="w"> </span>TCP.hmm<span class="w">  </span>/home/hanyapeng/prot/Gma.prot<span class="w"> </span>&gt;<span class="w"> </span>Gma.hmm<span class="w"> </span><span class="p">&amp;</span>
</span><span id="__span-3-2">less<span class="w"> </span>-S<span class="w"> </span>Gma.hmm<span class="w"> </span><span class="p">|</span>awk<span class="w"> </span><span class="s1">&#39;{if ($1~/e-/) print $9}&#39;</span><span class="w"> </span>&gt;<span class="w"> </span>list<span class="w">     </span><span class="c1">#if第一行匹配到e-，则调出第9行#  </span>
</span><span id="__span-3-3">
</span><span id="__span-3-4">nohup<span class="w"> </span>hmmsearch<span class="w"> </span>--domtblout<span class="w"> </span>hmmresult.txt<span class="w"> </span>-E<span class="w"> </span>1e-5<span class="w"> </span>TCP.hmm<span class="w">  </span>/home/hanyapeng/prot/Gma.prot<span class="w">  </span><span class="p">&amp;</span>
</span><span id="__span-3-5">less<span class="w"> </span>-S<span class="w"> </span>hmmresult.txt<span class="w"> </span><span class="p">|</span>awk<span class="w"> </span><span class="s1">&#39;{ if ($1~/e-/) print$1&quot;\t&quot;$20&quot;\t&quot;$21}&#39;</span><span class="p">|</span>less<span class="w"> </span>-S<span class="w"> </span>&gt;<span class="w"> </span>GmaTCPdomain.list<span class="w">    </span><span class="c1"># 提取第1.20.21列为坐标</span>
</span><span id="__span-3-6">
</span><span id="__span-3-7">perl<span class="w"> </span>/home/lixiangyong/software/getseq.pl<span class="w"> </span>Osa.list<span class="w"> </span>/home/hanyapeng/prot/Osa.prot<span class="w"> </span>&gt;<span class="w"> </span>Osa.fa
</span><span id="__span-3-8">perl<span class="w"> </span>/home/lixiangyong/software/get_domain.pl<span class="w">  </span>/home/hanyapeng/prot/Osa.prot<span class="w">  </span>OsaTCPdomain.list<span class="w"> </span>&gt;<span class="w"> </span>OsaTCPdomain.fa<span class="w">  </span><span class="c1">#提取domain</span>
</span><span id="__span-3-9">
</span><span id="__span-3-10">perl<span class="w"> </span>/home/lixiangyong/software/getseq.pl<span class="w"> </span>Chr13<span class="w"> </span>Gmax275.fa<span class="w"> </span>&gt;<span class="w"> </span>Chr13.fa<span class="w">  </span>
</span><span id="__span-3-11">perl<span class="w"> </span>/home/hanyapeng/comm/getpromoter.pl<span class="w"> </span>/data/plant_genome/Glyma2.0/Gmax275.fa<span class="w"> </span>Q1.txt<span class="w"> </span>&gt;<span class="w"> </span><span class="m">1</span><span class="w">  </span><span class="c1"># 提取启动子序列 Q1为启动子坐标</span>
</span><span id="__span-3-12">cat<span class="w"> </span>AT.fa<span class="w"> </span>Lj.fa<span class="w"> </span>Osa.fa<span class="w"> </span>&gt;<span class="w"> </span>ATLjOsa.fa
</span></code></pre></div></td></tr></table></div>
<h4 id="6"><a class="toclink" href="../2018/08/01/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6%E5%B8%B8%E7%94%A8%E7%9A%84linux%E5%91%BD%E4%BB%A4.html#6">6、 修改环境变量</a></h4>
<div class="language-text highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">Text Only</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span>
<span class="normal">5</span>
<span class="normal">6</span></pre></div></td><td class="code"><div><pre><span></span><code><span id="__span-4-1">vi .bash_profile     # 或则 .bash_login   .profile   .bashrc 
</span><span id="__span-4-2">vi .bash_profile     # 添加环境变量 PATH=$PATH:$HOME/bin
</span><span id="__span-4-3">source .bash_profile  #   立刻执行更改
</span><span id="__span-4-4">
</span><span id="__span-4-5">nano .bashrc             #和vi类似的编辑器
</span><span id="__span-4-6">source .bashrc
</span></code></pre></div></td></tr></table></div>
<h4 id="7-prottest"><a class="toclink" href="../2018/08/01/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6%E5%B8%B8%E7%94%A8%E7%9A%84linux%E5%91%BD%E4%BB%A4.html#7-prottest">7、 prottest的使用</a></h4>
<div class="language-text highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">Text Only</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span></pre></div></td><td class="code"><div><pre><span></span><code><span id="__span-5-1">nohup java -jar prottest-3.4.jar -i AGO.phy -all-matrices -all-distributions  &gt; a.txt  &amp;  # 此.phy 文件为比对过的氨基酸序列文件
</span><span id="__span-5-2">
</span><span id="__span-5-3">nohup java -jar prottest-3.4.jar -i namefail.phy -o namefail.txt -S 1 -AIC -I -G -IG -F -ncat 8 -all &amp;
</span><span id="__span-5-4">
</span><span id="__span-5-5">nohup java -jar prottest-3.4.jar -i AGOdomain.phy -o AGOdomain.txt -all-matrices -all-distributions &amp;
</span><span id="__span-5-6">
</span><span id="__span-5-7">nohup java -jar jModelTest.jar -d example-data/aP6.fas -f -i -g 4 -s 11 -AIC -a &gt;a.txt  &amp;  此.fas文件为核苷酸序列文件
</span><span id="__span-5-8">
</span><span id="__span-5-9">PhyML-3.1_linux64 -i aP6.phy -d nt -n 1 -b 0 --run_id F81+I -m 000000 -f m -v e -c 1 --no_memory_check -o tlr -s BEST
</span><span id="__span-5-10">
</span><span id="__span-5-11">PhyML-3.1_linux64 -i aP6.phy -q -d aa -m JTT -c 4 -a e
</span></code></pre></div></td></tr></table></div>
<h4 id="8-shell"><a class="toclink" href="../2018/08/01/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6%E5%B8%B8%E7%94%A8%E7%9A%84linux%E5%91%BD%E4%BB%A4.html#8-shell">8、 shell 脚本知识</a></h4>
<div class="language-shell highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">Bash</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div><pre><span></span><code><span id="__span-6-1">!<span class="o">=</span><span class="w"> </span>不等于,如:if<span class="w"> </span><span class="o">[</span><span class="w"> </span><span class="s2">&quot;</span><span class="nv">$a</span><span class="s2">&quot;</span><span class="w"> </span>!<span class="o">=</span><span class="w"> </span><span class="s2">&quot;</span><span class="nv">$b</span><span class="s2">&quot;</span><span class="w"> </span><span class="o">]</span><span class="w"> </span>
</span></code></pre></div></td></tr></table></div>
<h4 id="9-awk"><a class="toclink" href="../2018/08/01/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6%E5%B8%B8%E7%94%A8%E7%9A%84linux%E5%91%BD%E4%BB%A4.html#9-awk">9、 awk的使用</a></h4>
<div class="language-shell highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">Bash</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span></pre></div></td><td class="code"><div><pre><span></span><code><span id="__span-7-1">less<span class="w"> </span>Y2.blast<span class="w"> </span><span class="p">|</span>awk<span class="w"> </span><span class="s1">&#39; {if ($11 ~ /0\.0/) print $0}&#39;</span><span class="w"> </span>&gt;<span class="w"> </span><span class="m">1</span><span class="w">      </span><span class="c1"># 每一行中，如果第11列=0.0，则输出全部  “~”是 =  ，$11 调取第11列 ， \ 反义符 </span>
</span><span id="__span-7-2">awk<span class="w"> </span><span class="s1">&#39;{line[NR]=$0}END{for(i=NR;i&gt;0;i--)print line[i]}&#39;</span><span class="w"> </span>sed.txt<span class="w">  </span><span class="c1"># 把sed.txt 中的内容按行倒着输出</span>
</span><span id="__span-7-3">less<span class="w"> </span>-S<span class="w"> </span>Gma3.fa<span class="w"> </span><span class="p">|</span>awk<span class="w"> </span><span class="s1">&#39;{if ($1~/&gt;/) print $head}&#39;</span><span class="w"> </span>&gt;<span class="w"> </span>a
</span><span id="__span-7-4">awk<span class="w"> </span><span class="s1">&#39;!seen[$0]++&#39;</span><span class="w"> </span>&lt;filename&gt;<span class="w">    </span><span class="c1"># awk 去除重复的行</span>
</span><span id="__span-7-5">less<span class="w"> </span>Step1.2.sh<span class="w"> </span><span class="p">|</span><span class="w"> </span>grep<span class="w"> </span><span class="s2">&quot;=&quot;</span><span class="w"> </span><span class="p">|</span><span class="w"> </span>cut<span class="w"> </span>-d<span class="w"> </span><span class="s2">&quot;=&quot;</span><span class="w"> </span>-f<span class="w"> </span><span class="m">1</span><span class="w"> </span><span class="p">|</span><span class="w"> </span>awk<span class="w"> </span><span class="s1">&#39;{printf $quot&quot; &quot;;$0}&#39;</span><span class="w"> </span><span class="c1">#将一列输出为一行</span>
</span><span id="__span-7-6">awk<span class="w"> </span><span class="s1">&#39;{ print $(NF-1)}&#39;</span><span class="w"> </span>file<span class="w">     </span><span class="c1"># 出输出倒数第二列</span>
</span><span id="__span-7-7">
</span><span id="__span-7-8">less<span class="w"> </span>file<span class="w"> </span><span class="p">|</span><span class="w">  </span>awk<span class="w">  </span><span class="s1">&#39;{split($1,a,&quot;:&quot;);print a[1]&quot;-&quot;a[3]&quot;\t&quot;$5}&#39;</span><span class="w"> </span>&gt;<span class="w"> </span>our.w.d<span class="w">    </span><span class="c1"># 对awk截取字符串，file中的数据格式为:`ACAN:ENST00000561243:p.D1390E The mutant peptide: APGVEEISGLP is found in wild type sequence`， [参考](https://www.cnblogs.com/sunada2005/p/3493941.html)</span>
</span><span id="__span-7-9">
</span><span id="__span-7-10">awk<span class="w"> </span><span class="s1">&#39;BEGIN{total=0}{total+=$1}END{print total}&#39;</span><span class="w"> </span>file<span class="w">   </span><span class="c1"># 对文件第一列求和</span>
</span></code></pre></div></td></tr></table></div>
<h4 id="10-sed"><a class="toclink" href="../2018/08/01/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6%E5%B8%B8%E7%94%A8%E7%9A%84linux%E5%91%BD%E4%BB%A4.html#10-sed">10、 sed的使用</a></h4>
<div class="language-shell highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">Bash</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span></pre></div></td><td class="code"><div><pre><span></span><code><span id="__span-8-1">sed<span class="w"> </span>-i<span class="w"> </span><span class="s1">&#39;s/Gma/Glyma/g&#39;</span><span class="w"> </span>Glyma.fa
</span><span id="__span-8-2">sed<span class="w"> </span>-i<span class="w"> </span><span class="s1">&#39;/^$/d&#39;</span><span class="w"> </span>2_domain.txt<span class="w">  </span><span class="c1"># 删除文件空行</span>
</span><span id="__span-8-3">sed<span class="w"> </span>-i<span class="w"> </span><span class="s1">&#39;/#/d&#39;</span><span class="w"> </span>AtTCP.txt<span class="w">      </span><span class="c1"># 删除有\#的行</span>
</span><span id="__span-8-4">sed<span class="w"> </span>-i<span class="w"> </span>s/^M//<span class="w"> </span>AAt.list.txt<span class="w">   </span><span class="c1"># 此处^M必需手动输入！！</span>
</span><span id="__span-8-5">sed<span class="w"> </span>-i<span class="w"> </span>/^$/d<span class="w"> </span>file<span class="w">     </span><span class="c1">#删除空行   d 指删除含有空行符的这一行</span>
</span><span id="__span-8-6">sed<span class="w"> </span>-i<span class="w"> </span>s/tr.*//g<span class="w"> </span>LjTCP.fa<span class="w">  </span><span class="c1">#  去除无用的注释    :%s/|.*.\[//g  #去除从|开始，到[结束内的所有字符，包括[#</span>
</span><span id="__span-8-7">
</span><span id="__span-8-8">sed<span class="w"> </span><span class="s1">&#39;s/^/aaa/g&#39;</span><span class="w"> </span>filename.txt<span class="w"> </span>&gt;<span class="w"> </span>outputfile.txt<span class="w">  </span><span class="c1"># 在文件中的每一行之前加上一个字符串，比如：aaa    行尾将^是 $ </span>
</span><span id="__span-8-9">sed<span class="w"> </span>-i/<span class="se">\&gt;</span>Chr17/d<span class="w"> </span>Chr17.fa<span class="w">   </span><span class="c1"># 删除Chr17.fa 中的Chr字符</span>
</span><span id="__span-8-10">sed<span class="w"> </span>-i<span class="w"> </span><span class="s2">&quot;s/zhangsan/lisi/g&quot;</span><span class="w"> </span><span class="sb">`</span>grep<span class="w"> </span>zhangsan<span class="w"> </span>-rl<span class="w"> </span>modules<span class="sb">`</span><span class="w"> </span><span class="c1">#批量替换，讲文件夹modules中所有文件执行sed -i       &quot;s/zhangsan/lisi/g&quot; 该符号`不是点，是Esc下面的符号 </span>
</span><span id="__span-8-11">sed<span class="w"> </span>-i<span class="w"> </span><span class="s1">&#39;1iNew Top Line&#39;</span><span class="w"> </span>aa.txt<span class="w">    </span><span class="c1">#  sed 在句首添加一行 New Top Line</span>
</span></code></pre></div></td></tr></table></div>
<h4 id="11-linux"><a class="toclink" href="../2018/08/01/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6%E5%B8%B8%E7%94%A8%E7%9A%84linux%E5%91%BD%E4%BB%A4.html#11-linux">11、 linux 命令</a></h4>
<div class="language-shell highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">Bash</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span></pre></div></td><td class="code"><div><pre><span></span><code><span id="__span-9-1">date<span class="w">   </span><span class="c1">#  查看系统日期</span>
</span><span id="__span-9-2">cal<span class="w"> </span><span class="m">1</span><span class="w"> </span><span class="m">2016</span><span class="w">   </span><span class="c1"># 查看2016年1月的日历</span>
</span><span id="__span-9-3">df<span class="w">     </span><span class="c1"># 查看磁盘驱动器的可用空间  </span>
</span><span id="__span-9-4">free<span class="w">   </span><span class="c1"># 显示可用内存</span>
</span><span id="__span-9-5">.<span class="w">     </span><span class="c1"># 工作目录     </span>
</span><span id="__span-9-6">..<span class="w">   </span><span class="c1"># 工作目录的父目录   如：  cd ..   返回上一级    cd  ./software/  </span>
</span><span id="__span-9-7">file<span class="w"> </span>Gma1.fa<span class="w">   </span><span class="c1"># 显示  Gma1.fa: ASCII text  该文件为文本文档ASCII</span>
</span><span id="__span-9-8"><span class="nb">alias</span><span class="w"> </span><span class="nv">foo</span><span class="o">=</span><span class="s1">&#39;cd /usr; ls; cd -&#39;</span><span class="w">   </span><span class="c1"># 自己创建命令   unalias foo 删除</span>
</span><span id="__span-9-9">gedit<span class="w"> </span>Gma1.fa<span class="w">   </span><span class="c1"># 图形编辑器编辑Gma1.fa </span>
</span><span id="__span-9-10">
</span><span id="__span-9-11">rm<span class="w"> </span>--<span class="w"> </span>-foo<span class="w">   </span><span class="c1"># 要删除第一个字符为‘-’的文件 (例如‘-foo’)</span>
</span><span id="__span-9-12">rm<span class="w"> </span>./-foo<span class="w">   </span><span class="c1"># 要删除第一个字符为‘-’的文件 (例如‘-foo’)</span>
</span><span id="__span-9-13">
</span><span id="__span-9-14">chmod<span class="w"> </span><span class="m">755</span><span class="w"> </span>lixiangyong/<span class="w">  </span><span class="c1"># 更改目录权限为可读</span>
</span><span id="__span-9-15">
</span><span id="__span-9-16">wc<span class="w"> </span>-l<span class="w"> </span>filename<span class="w"> </span><span class="c1"># 就是查看文件里有多少行</span>
</span><span id="__span-9-17">wc<span class="w"> </span>-w<span class="w"> </span>filename<span class="w"> </span><span class="c1"># 看文件里有多少个word。</span>
</span><span id="__span-9-18">wc<span class="w"> </span>-L<span class="w"> </span>filename<span class="w"> </span><span class="c1"># 文件里最长的那一行是多少个字</span>
</span><span id="__span-9-19">wc<span class="w"> </span>-c<span class="w">  </span>filename<span class="w">   </span><span class="c1"># 统计字节数</span>
</span><span id="__span-9-20">
</span><span id="__span-9-21">cat<span class="w"> </span>Y.list<span class="w"> </span><span class="p">|</span><span class="w"> </span>sort<span class="w"> </span>-u<span class="w"> </span>&gt;<span class="w"> </span>YRedundancy.list<span class="w">    </span><span class="c1"># 去除文件Y.list中的重复的行命名为YR.list</span>
</span><span id="__span-9-22">less<span class="w"> </span>-S<span class="w"> </span>AGO.fa<span class="w"> </span><span class="p">|</span>grep<span class="w"> </span><span class="s2">&quot;&gt;&quot;</span><span class="w"> </span><span class="p">|</span>wc<span class="w"> </span>-l<span class="w">  </span><span class="c1"># 检索&gt;有多少个  同  grep -c &quot;&gt;&quot; AGO.fa</span>
</span><span id="__span-9-23"><span class="k">for</span><span class="w"> </span>tar<span class="w"> </span><span class="k">in</span><span class="w"> </span>*.tar.gz<span class="p">;</span><span class="w">  </span><span class="k">do</span><span class="w"> </span>tar<span class="w"> </span>xvf<span class="w"> </span><span class="nv">$tar</span><span class="p">;</span><span class="w"> </span><span class="k">done</span><span class="w">    </span><span class="c1">#批量解压文件#</span>
</span><span id="__span-9-24">split<span class="w"> </span>-l<span class="w"> </span><span class="m">10</span><span class="w"> </span>splitfile.txt<span class="w"> </span>D<span class="w">  </span><span class="c1"># 将文件按每10行分割，新文件命名为 Daa Dab#</span>
</span><span id="__span-9-25">rename<span class="w"> </span>linux<span class="w"> </span>unix<span class="w"> </span>linux*<span class="w">    </span><span class="c1"># 文件批量  </span>
</span><span id="__span-9-26">find<span class="w"> </span>.<span class="w"> </span>-name<span class="w"> </span><span class="s1">&#39;*.txt&#39;</span><span class="w"> </span><span class="p">|</span><span class="w"> </span>xargs<span class="w"> </span>perl<span class="w"> </span>-pi<span class="w"> </span>-e<span class="w"> </span><span class="s1">&#39;s/aaa/bbb/g&#39;</span><span class="w">  </span><span class="c1"># 遍历整个文件夹，把所有txt文件中的aaa替换成bbb</span>
</span><span id="__span-9-27">mkdir<span class="w"> </span><span class="k">$(</span>date<span class="w"> </span>+%F<span class="k">)</span><span class="w">  </span><span class="c1">#创建一个以日期为名的文件夹</span>
</span><span id="__span-9-28">touch<span class="w"> </span><span class="o">{</span><span class="m">1</span>..9<span class="o">}</span>.txt<span class="w">   </span><span class="c1"># 批量创建文件</span>
</span><span id="__span-9-29">
</span><span id="__span-9-30">ls<span class="w"> </span>-F<span class="w"> </span><span class="p">|</span><span class="w"> </span>grep<span class="w"> </span><span class="s1">&#39;/$&#39;</span><span class="w">         </span><span class="c1">#  查看当前目录下的文件夹 </span>
</span><span id="__span-9-31"><span class="k">for</span><span class="w"> </span>i<span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="sb">`</span>ls<span class="w"> </span>-F<span class="w"> </span><span class="p">|</span><span class="w"> </span>grep<span class="w"> </span><span class="s1">&#39;/$&#39;</span><span class="sb">`</span><span class="p">;</span><span class="k">do</span><span class="w"> </span>sudo<span class="w"> </span>du<span class="w"> </span>-sh<span class="w"> </span><span class="nv">$i</span><span class="p">;</span><span class="k">done</span><span class="w">   </span><span class="c1">#查看当前目录下，每个子目录及其子目录的总大小</span>
</span><span id="__span-9-32">
</span><span id="__span-9-33">rename<span class="w"> </span><span class="s2">&quot;s/AA/aa/&quot;</span><span class="w"> </span>*
</span><span id="__span-9-34">rename<span class="w"> </span>.fas<span class="w"> </span>.fa<span class="w"> </span>*.fas<span class="w">    </span><span class="c1"># CentOS 下可用</span>
</span><span id="__span-9-35">
</span><span id="__span-9-36">iostat<span class="w"> </span>-d<span class="w"> </span>-k<span class="w"> </span><span class="m">1</span><span class="w"> </span>10可以看到每个盘的读写速度
</span><span id="__span-9-37">
</span><span id="__span-9-38">find<span class="w"> </span>/home/lxy<span class="w"> </span>-type<span class="w"> </span>f<span class="w"> </span>-size<span class="w"> </span>+100M<span class="w">    </span><span class="c1">#列出文件夹 /home/lxy 下大于100M的文件</span>
</span><span id="__span-9-39">find<span class="w"> </span>.<span class="w"> </span>-type<span class="w"> </span>l<span class="w"> </span>-exec<span class="w"> </span>rm<span class="w">  </span><span class="o">{}</span><span class="w"> </span><span class="se">\;</span><span class="w">  </span><span class="c1">#列出当前路径 .  下的link类型的文件(type</span>
</span><span id="__span-9-40"><span class="w"> </span>l<span class="o">)</span><span class="w"> </span>并删除
</span></code></pre></div></td></tr></table></div>
<h4 id="12-perl"><a class="toclink" href="../2018/08/01/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6%E5%B8%B8%E7%94%A8%E7%9A%84linux%E5%91%BD%E4%BB%A4.html#12-perl">12、 单行 perl</a></h4>
<div class="language-text highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">Text Only</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div><pre><span></span><code><span id="__span-10-1">perl -e &quot;print lc($a)&quot;    # Perl 大小写转换
</span></code></pre></div></td></tr></table></div>
<h4 id="13-r"><a class="toclink" href="../2018/08/01/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6%E5%B8%B8%E7%94%A8%E7%9A%84linux%E5%91%BD%E4%BB%A4.html#13-r">13、 R 命令行</a></h4>
<div class="language-text highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">Text Only</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div><pre><span></span><code><span id="__span-11-1">Rscript /home/muscle/test.r  # linux 上运行r脚本
</span></code></pre></div></td></tr></table></div>
<h4 id="14-kaks"><a class="toclink" href="../2018/08/01/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6%E5%B8%B8%E7%94%A8%E7%9A%84linux%E5%91%BD%E4%BB%A4.html#14-kaks">14、 kaks计算</a></h4>
<div class="language-text highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">Text Only</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div><pre><span></span><code><span id="__span-12-1">perl pal2nal.perl aligned.protein.fas original.cds.fas -output fasta &gt; aligned.cds.fas
</span><span id="__span-12-2">perl ../software/pal2nal.pl 2.fas 2.cds -output paml &gt; 2.nuc    #kaks计算
</span></code></pre></div></td></tr></table></div>
<h4 id="15-gffgtf"><a class="toclink" href="../2018/08/01/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6%E5%B8%B8%E7%94%A8%E7%9A%84linux%E5%91%BD%E4%BB%A4.html#15-gffgtf">15、 gff和gtf之间的格式转化</a></h4>
<div class="language-text highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">Text Only</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div><pre><span></span><code><span id="__span-13-1">gffread my.gff3 -T -o my.gtf # gff2gtf
</span><span id="__span-13-2">gffread merged.gtf -o- &gt; merged.gff3  # gtf2gff
</span></code></pre></div></td></tr></table></div>
<h4 id="16-vim"><a class="toclink" href="../2018/08/01/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6%E5%B8%B8%E7%94%A8%E7%9A%84linux%E5%91%BD%E4%BB%A4.html#16-vim">16、 vim 的使用</a></h4>
<div class="language-text highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">Text Only</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span>
<span class="normal">5</span>
<span class="normal">6</span>
<span class="normal">7</span>
<span class="normal">8</span>
<span class="normal">9</span></pre></div></td><td class="code"><div><pre><span></span><code><span id="__span-14-1">:12,23s/aa/bb/g     # 将从12 行到23 行中出现的所有包含aa 的字符串中的aa 替换为bb
</span><span id="__span-14-2">:s/\&lt;aa\&gt;/bb/g      # 将光标所在行出现的所有aa 替换为bb, 仅替换aa 这个单词
</span><span id="__span-14-3">:12,23s/^/#/        # 将从12 行到23 行的行首加入# 字符
</span><span id="__span-14-4">
</span><span id="__span-14-5">:10,15s#^#\#\##g          #  vim的 V模式下，下注释x  10 -15  行，##注释
</span><span id="__span-14-6">
</span><span id="__span-14-7">Esc shift+ : wq   enter   =  Esc shift+ : x  enter       #保存并退出
</span><span id="__span-14-8">Esc shift+ : q！  enter                                  #不保存强制退出 
</span><span id="__span-14-9">vi foo.txt ls-output.txt    # 同时vi编辑多个文件  ：n   ：N   ：buffer 2 来回切换     也可以先打开一个，在用：e ls-output.txt
</span></code></pre></div></td></tr></table></div>
<h4 id="17"><a class="toclink" href="../2018/08/01/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6%E5%B8%B8%E7%94%A8%E7%9A%84linux%E5%91%BD%E4%BB%A4.html#17">17、 数据存放</a></h4>
<p>/data/public/hanyapeng/Gm/Gma.collinear.groups      # 大豆中所有旁系同源基因<br />
/data/public/wanglei/inparanoid4.0/Soybeanparalog2.txt     # 大豆中所有旁系同源基因</p>
<h3 id="_1"><a class="toclink" href="../2018/08/01/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6%E5%B8%B8%E7%94%A8%E7%9A%84linux%E5%91%BD%E4%BB%A4.html#_1">推荐连接</a></h3>
<ul>
<li><a href="https://github.com/stephenturner/oneliners">Bioinformatics one-liners</a></li>
</ul>
    
  </div>
</article>
      
      
        
          



<nav class="md-pagination">
  
</nav>
        
      
    </div>
  </div>

          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg>
  回到页面顶部
</button>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Copyright © 2024 biolxy
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
        <div class="md-social">
  
    
    
    
    
    <a href="https://github.com/biolxy" target="_blank" rel="noopener" title="GitHub | biolxy" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8M97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.tabs", "navigation.top", "navigation.indexes", "navigation.expand", "search.suggest", "search.highlight", "content.code.copy", "content.action.edit"], "search": "../../assets/javascripts/workers/search.d50fe291.min.js", "tags": null, "translations": {"clipboard.copied": "\u5df2\u590d\u5236", "clipboard.copy": "\u590d\u5236", "search.result.more.one": "\u5728\u8be5\u9875\u4e0a\u8fd8\u6709 1 \u4e2a\u7b26\u5408\u6761\u4ef6\u7684\u7ed3\u679c", "search.result.more.other": "\u5728\u8be5\u9875\u4e0a\u8fd8\u6709 # \u4e2a\u7b26\u5408\u6761\u4ef6\u7684\u7ed3\u679c", "search.result.none": "\u6ca1\u6709\u627e\u5230\u7b26\u5408\u6761\u4ef6\u7684\u7ed3\u679c", "search.result.one": "\u627e\u5230 1 \u4e2a\u7b26\u5408\u6761\u4ef6\u7684\u7ed3\u679c", "search.result.other": "# \u4e2a\u7b26\u5408\u6761\u4ef6\u7684\u7ed3\u679c", "search.result.placeholder": "\u952e\u5165\u4ee5\u5f00\u59cb\u641c\u7d22", "search.result.term.missing": "\u7f3a\u5c11", "select.version": "\u9009\u62e9\u5f53\u524d\u7248\u672c"}, "version": null}</script>
    
    
      <script src="../../assets/javascripts/bundle.13a4f30d.min.js"></script>
      
        <script src="../../javascripts/config.js"></script>
      
        <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
      
    
  </body>
</html>